File: complete_search/131_palindrome_partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        parts = []
        def isPalindrome(s, l, r):
            while l <= r:
                if s[l] != s[r]:
                    return False
                l, r = l+1, r -1
            return True

        def dfs(i):
            if i == len(s):
                res.append(parts.copy())
                return
            for j in range(i, len(s)):
                print(s[i: j+1])
                if isPalindrome(s, i, j):
                    parts.append(s[i: j + 1])
                    dfs(j + 1)
                    parts.pop()
        
        dfs(0)
        return res

    
        

File: backtracking/131_palindrome_partitioning.java
class Solution {
    int n = 0;
    String s;

    List<List<String>> res = new ArrayList<List<String>>();
    
    public boolean isPalindrome(int left, int right) {
        while (left < right) {
            if (this.s.charAt(left++) != this.s.charAt(right--)) {
                return false;
            }
        }
        return true;
    }

    public void dfs(int start, List<String> curr) {
        if (start >= n) {
            res.add(new ArrayList<String>(curr));
        }
        for (int end = start; end < n; end++) {
            if (isPalindrome(start, end)) {
                curr.add(this.s.substring(start, end + 1));
                dfs(end + 1, curr);
                curr.remove(curr.size() - 1);
            }
        }

    }

    public List<List<String>> partition(String s) {
        n = s.length();
        this.s = s;
        
        dfs(0, new ArrayList<String>());        
        return res;
    }
}